File: app/src/main/java/com/practice/lc/recursion/java/Leetcode394.java
package com.practice.lc.recursion.java;

import java.util.ArrayDeque;

public class Leetcode394 {
    public record State(StringBuilder sb, int repeat) {}
    public String decodeString(String s) {
        final ArrayDeque<State> stack = new ArrayDeque<>();
        StringBuilder sb = new StringBuilder();
        StringBuilder repeat = new StringBuilder();

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) >= 'a' && s.charAt(i) <= 'z') {
                sb.append(s.charAt(i));
            } else if (s.charAt(i) >= '0' && s.charAt(i) <= '9') {
                repeat.append(s.charAt(i));
            } else if (s.charAt(i) == '[') {
                final State state = new State(sb, Integer.parseInt(repeat.toString()));
                stack.addLast(state);
                sb = new StringBuilder();
                repeat = new StringBuilder();
            } else if (s.charAt(i) == ']') {
                final State state = stack.removeLast();
                for (int j = 0; j < state.repeat; j++) {
                    state.sb.append(sb);
                }
                sb = state.sb;
            }
        }

        return sb.toString();
    }
}

File: app/src/main/java/com/practice/lc/Leetcode394.kt
package com.practice.lc

import kotlin.text.buildString
import java.lang.StringBuilder

class Leetcode394 {
    public fun decodeString(s: String): String {
        val stack = ArrayDeque<Pair<StringBuilder, Int>>()
        var decodeStrSoFar = StringBuilder()
        var digits = StringBuilder()

        for (c in s) {
            if (c in 'a'..'z') {
                decodeStrSoFar.append(c)
                continue
            }

            if (c in '0'..'9') {
                digits.append(c)
                continue
            }

            if (c == '[') {
                stack.addLast(Pair<StringBuilder, Int>(decodeStrSoFar, digits.toString().toInt()))
                decodeStrSoFar = StringBuilder()
                digits = StringBuilder()
                continue
            }

            if (c == ']') {
                val (savedSb, savedInt) = stack.removeLast()
                val tempStr = buildString {
                    for (i in 0 until savedInt) {
                        append(decodeStrSoFar)
                    }
                }
                savedSb.append(tempStr)
                decodeStrSoFar = savedSb
            }
        }

        return decodeStrSoFar.toString()
    }
}